src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -389,45 +389,65 @@ SegmentCleaner::gc_reclaim_space_ret SegmentCleaner::gc_reclaim_space()
           "reclaim_space",
           [this, &extents](auto& t)
         {
-          return trans_intr::do_for_each(
-              extents,
-              [this, &t](auto &extent) {
-	    auto& addr = extent.first;
-	    auto& info = extent.second;
-            logger().debug(
-              "SegmentCleaner::gc_reclaim_space: checking extent {}",
-              info);
-            return ecb->get_extent_if_live(
-              t,
-              info.type,
-              addr,
-              info.addr,
-              info.len
-            ).si_then([addr=addr, &t, this, &info](CachedExtentRef ext) {
-              if (!ext) {
-                logger().debug(
-                  "SegmentCleaner::gc_reclaim_space: addr {} dead, skipping",
-                  addr);
-                return ExtentCallbackInterface::rewrite_extent_iertr::now();
-              } else {
-                logger().debug(
-                  "SegmentCleaner::gc_reclaim_space: addr {} alive, gc'ing {}",
-                  addr,
-                  *ext);
-		assert(info.last_modified);
-		if (ext->get_last_modified() == ceph::real_clock::time_point()) {
-		  ext->set_last_modified(ceph::real_clock::duration(info.last_modified));
+	  return seastar::do_with(
+	    // TODO: this is a copy, to make sure we can re-add all extents to
+	    // 	     the transaction when there's transaction conflicts, need make
+	    // 	     some optimization
+	    extents,
+	    [this, &t](auto& extents) {
+	    return trans_intr::repeat(
+		[this, &t, &extents]() mutable {
+	      if (extents.empty()) {
+		return ExtentCallbackInterface::rewrite_extent_iertr::make_ready_future<
+		  seastar::stop_iteration>(seastar::stop_iteration::yes);
+	      }
+	      auto& extent = extents.top();
+	      auto& addr = extent.first;
+	      auto& info = extent.second;
+	      logger().debug(
+		"SegmentCleaner::gc_reclaim_space: checking extent {}, {} extents left",
+		info, extents.size());
+	      return ecb->get_extent_if_live(
+		t,
+		info.type,
+		addr,
+		info.addr,
+		info.len
+	      ).si_then([addr=addr, &t, this, &info](CachedExtentRef ext) {
+		if (!ext) {
+		  logger().debug(
+		    "SegmentCleaner::gc_reclaim_space: addr {} dead, skipping",
+		    addr);
+		  return ExtentCallbackInterface::rewrite_extent_iertr::now();
+		} else {
+		  logger().debug(
+		    "SegmentCleaner::gc_reclaim_space: addr {} alive, gc'ing {}",
+		    addr,
+		    *ext);
+		  assert(info.last_modified);
+		  if (ext->get_last_modified() == ceph::real_clock::time_point()) {
+		    ext->set_last_modified(ceph::real_clock::duration(info.last_modified));
+		  } else {
+		    assert(info.last_modified <=
+		      ext->get_last_modified().time_since_epoch().count());
+		  }
+		  ext->set_last_modified(ceph::real_clock::now());
```

Comment:
Ah, this is a little bit tricky. We have to reset extents' last_modified to now, because, otherwise, as the time goes on, long live extents will always win and gets gc'd, which means there is a time point after which only segments holding long live extents will get gc'd even if there are lots of segments emptied by short lived extents overwrites. This `ext->set_last_modified(ceph::real_clock::now());` is that reset.

Reply:
> there is a time point after which only segments holding long live extents will get gc'd even if there are lots of segments emptied by short lived extents overwrites.

I don't quite follow, if we can gather all the long live extents into the same segment during reclaim, the `util` of the cold segment should drop from 1 at much slower speed than the hotter segments. This means that the benefit value `(1 - util) * age / (1 + util)` of the cold segments will keep low (~0) for a long time until its `util` grows up to be chosen by the cleaner.

I still tend to believe we should not reset the age of the extent unless we actually modified it, otherwise we are losing the only information to identify the aged extents.

Reply:
> I don't quite follow, if we can gather all the extents of the similar age into the same segment during reclaim, the `util` of the cold segment should drop from 1 at much slower speed than the hotter segments. This means that the benefit value `(1 - util) * age / (1 + util)` of the cold segments will keep low (~0) for a long time until its `util` grows up to be chosen by the cleaner.

Actually, not that long. In our environment, after about 24 hours of testing, long lived extents became the "forever" winner.

> I still tend to believe we should not reset the age of the extent unless we actually modified it, otherwise we are losing the only information to identify the aged extents.

We did this after encountering the above scenario.

Reply:
Instead of resetting the age information, another option might be to tame the benefit calculation.

The original one may not be the best fit as (1 + util) assumes that the read overhead is equal to the write overhead. In our case, the age parameter proves to introduce too large impact on the final benefit value. Or, it may because we used the last-last modified-time instead of the mean-last-modified-time of the segment.

Reply:
> Or, it may because we used the last-last modified-time instead of the mean-last-modified-time of the segment.

I don't get it, even if we use mean-last-modfied-time instead, it will at most make the forever winner appear later, this can't solve the issue.

Reply:
> the age parameter proves to introduce too large impact on the final benefit value

How about `(1 - util) * log(age) / (1 + util)` ?

Reply:
> How about `(1 - util) * log(age + 1) / (1 + util)` ?

I think this could push back the time that cold extents get reclaimed, which means the humps in the bimodal distribution of sprite LFS segment utilization would be much closer. According to the paper, most segments were either 20% or 80% utilized. If we use `log(age + 1)`, those numbers could be closer, like 40% to 60%. Right now, we can't confirm this for the reason discussed [here](https://github.com/ceph/ceph/pull/43474#issuecomment-985278502).

Reply:
I'm with @cyx1231st, I really don't think it makes sense to be resetting the modification time.  Can you explain in more detail what problem you encountered?

Reply:
> I'm with @cyx1231st, I really don't think it makes sense to be resetting the modification time. Can you explain in more detail what problem you encountered?

The gc benefit is `(1 - util) * age / (1 + util)` as stated in the spritelfs paper. The issue is that the `util` drops slower as time goes by, while the `age`'s increase speed keeps steady, so there comes a time when `cold extents`' gc benefit always overrun other extents and the gc will always reclaim those cold extents even if they can hardly provide any new space.

In my tests, this time in point came about 24 hours after the tests started. So I thought there should be a boundary for the `age` and chose the time it gets reclaimed as that boundary. Actually, I thought that's what the paper implied as it didn't explicitly state what the `age` is while its tests run successfully. I mean there has to be some other things that made its tests succeeded and the paper didn't mention . Am I right?

Reply:
@cyx1231st @athanatos Actually, my first thought was also correcting the equation for calculating the gc benefit. But spritelfs' equation was a successful one as long as the spritelfs paper didn't lie. So I thought there must be something else that needs to be corrected. Am I right?

Reply:
I think what I chose was to consider the `age` as the age since the last time the extent got rewritten by gc. Does this make sense?

Reply:
We probably shouldn't assume that the paper's heuristics and assumptions hold true for ceph.  We have a very different write pattern and aren't actually writing to a spinning disk.  Let's start with trying to maintain actual modification times and adapt the gc benefit calculation to use them appropriately.  Is there a relatively simple gc benefit calculation we can start with?

Reply:
Before we can find out a proper strategy, probably worth a dedicated tool to profile the cleaning process and reveal the above issues consistently. So that we can have a baseline to start with.

Reply:
If all agree, I think I can take some time to formulate this issue, so that we can discuss it precisely.

Reply:
@athanatos @cyx1231st

I think maybe we can solve this issue this way: add a new field `last_rewritten` to CachedExtent besides the `last_modified` field, and calculate the `age` like this: `age = chrono::now() - std::max(extent->last_rewitten, extent->last_modified)`. 

With this approach, we can fix the problem mentioned above without reseting `extent->last_modified`. And if this calculation turns out to be premature, all we have to do in the future is modifying the gc benefit calculation.

Generally speaking, we just feed information to the `SegmentCleaner::calc_gc_benefit_cost()` as much as possible, and let that method decide what to do next.

How do you like it?